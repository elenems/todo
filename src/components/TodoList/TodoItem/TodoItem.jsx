import React from "react";
import IconButton from "@material-ui/core/IconButton";
import { makeStyles } from "@material-ui/core/styles";
import { MdRemoveCircleOutline } from "react-icons/md";
import { MdEdit } from "react-icons/md";
import Box from "@material-ui/core/Box";
import { connect } from "react-redux";
import * as Actions from "../../../store/actions/todoActions";
import Tooltip from "@material-ui/core/Tooltip";
import Typography from "@material-ui/core/Typography";
import { withRouter } from "react-router-dom";
const useStyles = makeStyles({
  root: {
    margin: "0px 0px 10px 0px",
    padding: "12px",
    fontSize: "26px",
    border: "none"
  }
});

function TodoItem(props) {
  const todo = props.todo;
  const classes = useStyles();

  const moveToEditTodo = () => {
    const todoID = props.todo.id;
    props.history.push(`/edit/${todoID}`);
  };

  const removeTodo = e => {
    const id = todo.id;
    const element = e.target.closest(".todo-item");
    element.className += ' hide';
    setTimeout(() => {
      props.removeTodo(id);
    }, 400);
  };

  const toggleTodoStatus = e => {
    props.toggleTodoStatus(props.todo.id);
    const element = e.target;
    const textDecoration =
      element.style.textDecoration === "line-through" ? "none" : "line-through";
    element.style.textDecoration = textDecoration;
  };

  const textStyle = props.todo.isChecked
    ? { textDecoration: "line-through" }
    : { textDecoration: "none" };

  return (
    <Box
      className="todo-item reveal"
      display="flex"
      alignItems="baseline"
      justifyContent="space-between"
    >
      <Typography
        style={textStyle}
        onClick={toggleTodoStatus}
        component="p"
        variant="h5"
      >
        {todo.title}
      </Typography>
      <Box>
        <Tooltip title="Edit">
          <IconButton
            onClick={moveToEditTodo}
            color="primary"
            aria-label="edit"
            classes={{ root: classes.root }}
          >
            <MdEdit />
          </IconButton>
        </Tooltip>
        <Tooltip title="Delete">
          <IconButton
            onClick={removeTodo}
            color="secondary"
            aria-label="delete"
            classes={{ root: classes.root }}
          >
            <MdRemoveCircleOutline />
          </IconButton>
        </Tooltip>
      </Box>
    </Box>
  );
}

const mapDispatchToProps = dispatch => {
  return {
    removeTodo: id => dispatch(Actions.removeTodo(id)),
    toggleTodoStatus: id => dispatch(Actions.toggleTodoStatus(id))
  };
};

export default withRouter(
  connect(
    null,
    mapDispatchToProps
  )(TodoItem)
);
