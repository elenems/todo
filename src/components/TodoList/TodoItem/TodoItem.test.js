import { shallow } from "enzyme";
import React from "react";
import TodoItem from "../TodoItem";
import { TodoItemTest } from "./TodoItem.jsx";

describe("TodoItem", () => {
  let props;
  beforeAll(() => {
    const todo = { id: 1, title: "Title", isChecked: false };
    props = {
      toggleTodoStatus() {},
      removeTodo() {},
      todo,
      history: {
        push() {}
      }
    };
  });
  it("Matches snapshot", () => {
    const wrapper = shallow(<TodoItemTest {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("Can click moveToEditTodo", () => {
    const wrapper = shallow(<TodoItemTest {...props} />);
    wrapper.find("#editTodoRedirectButton").simulate("click");
    expect(wrapper.find("#editTodoRedirectButton").length).toBe(1);
  });

  it("Can click removeTodoButton", () => {
    const wrapper = shallow(<TodoItemTest {...props} />);
    const elem = document.createElement("div");
    wrapper.find("#removeTodoButton").simulate("click", {
      target: {
        closest() {
          return elem;
        }
      }
    });
    expect(wrapper.find("#removeTodoButton").length).toBe(1);
  });

  it("Can click todoStatusToggler", () => {
    const elem = document.createElement("p");
    const wrapper = shallow(<TodoItemTest {...props} />);
    wrapper.find("#todoStatusToggler").simulate("click", { target: elem });
    expect(wrapper.find("#todoStatusToggler").length).toBe(1);
  });

  it("Cals remove todo after delay", () => {
    jest.useFakeTimers();

    const wrapper = shallow(<TodoItemTest {...props} />);
    const elem = document.createElement("div");
    const mock = jest.fn();
    wrapper.find("#removeTodoButton").simulate("click", {
      target: {
        closest() {
          return elem;
        }
      }
    });
    jest.runAllTimers();
    expect(setTimeout).toHaveBeenCalledTimes(1);
  });

  it("S", () => {
    const wrapper = shallow(<TodoItem />);
    console.log(wrapper.props())
  });
});
