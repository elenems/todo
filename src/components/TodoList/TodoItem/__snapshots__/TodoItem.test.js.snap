// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem Matches snapshot 1`] = `
<Styled(MuiBox)
  alignItems="baseline"
  className="todo-item reveal"
  display="flex"
  justifyContent="space-between"
>
  <WithStyles(ForwardRef(Typography))
    component="p"
    id="todoStatusToggler"
    onClick={[Function]}
    style={
      Object {
        "textDecoration": "none",
      }
    }
    variant="h5"
  >
    Title
  </WithStyles(ForwardRef(Typography))>
  <Styled(MuiBox)>
    <WithStyles(ForwardRef(Tooltip))
      title="Edit"
    >
      <WithStyles(ForwardRef(IconButton))
        aria-label="edit"
        classes={
          Object {
            "root": "makeStyles-root-1",
          }
        }
        color="primary"
        id="editTodoRedirectButton"
        onClick={[Function]}
      >
        <MdEdit />
      </WithStyles(ForwardRef(IconButton))>
    </WithStyles(ForwardRef(Tooltip))>
    <WithStyles(ForwardRef(Tooltip))
      title="Delete"
    >
      <WithStyles(ForwardRef(IconButton))
        aria-label="delete"
        classes={
          Object {
            "root": "makeStyles-root-1",
          }
        }
        color="secondary"
        id="removeTodoButton"
        onClick={[Function]}
      >
        <MdRemoveCircleOutline />
      </WithStyles(ForwardRef(IconButton))>
    </WithStyles(ForwardRef(Tooltip))>
  </Styled(MuiBox)>
</Styled(MuiBox)>
`;
